File: apps/worker-media/src/queues.ts
/**
 * Queue names, the job envelope and the media payloads, frozen in
 * `docs/CONTRACTS.md` §3.
 *
 * A08 owns the canonical TypeScript copy (`apps/api/src/jobs/contracts/`), but a
 * worker cannot import across app boundaries, so the literal names live here and
 * `queues.test.ts` parses the API's file to prove the two have not drifted — the
 * same guard `apps/worker-ai/tests/test_queues.py` gives the Python side.
 *
 * `montaj` in the queue names is the engineering codename and is correct — the
 * brand rule (CONTRACTS §0) covers user-visible strings only.
 */

export const QUEUE_NAMES = [
  "media.probe",
  "media.proxy",
  "media.acquire",
  "media.clip",
  "ai.vad",
  "ai.transcribe",
  "ai.align",
  "ai.diarise",
  "ai.translate",
  "ai.transliterate",
  "ai.clean",
  "ai.pass",
  "ai.llm",
  "ai.highlights",
  "ai.faces",
  "render.video",
  "render.subtitle",
  "publish.dispatch",
  "publish.reconcile",
  "notify",
] as const;

export type QueueName = (typeof QUEUE_NAMES)[number];

/**
 * Queues this worker consumes.
 *
 * `media.clip` is in QUEUE_NAMES from REP-005 but is NOT here: registering a name
 * is a contract, consuming it is an implementation, and its processor arrives in
 * Wave 6. A job enqueued on it today would sit in Redis, which is why nothing
 * enqueues it until then.
 *
 * `media.acquire` IS consumed (REP-010), and since 2026-09-15 it is produced too:
 * creating a link-sourced run enqueues one. It stays gated on
 * `source_youtube_acquire`, which is seeded off, so a deployment that has not
 * enabled the flag still sees no acquisitions.
 */
export const MEDIA_PROBE_QUEUE = "media.probe" satisfies QueueName;
export const MEDIA_PROXY_QUEUE = "media.proxy" satisfies QueueName;
export const MEDIA_ACQUIRE_QUEUE = "media.acquire" satisfies QueueName;
export const MEDIA_CLIP_QUEUE = "media.clip" satisfies QueueName;

/** The media queues consumed by this worker. */
export const MEDIA_QUEUES = [
  MEDIA_ACQUIRE_QUEUE,
  MEDIA_PROBE_QUEUE,
  MEDIA_PROXY_QUEUE,
  MEDIA_CLIP_QUEUE,
] as const;

export type MediaQueue = (typeof MEDIA_QUEUES)[number];

/** Every job's `data`, identical across queues (CONTRACTS §3). */
export interface JobEnvelope<TPayload = unknown> {
  readonly jobId: string;
  readonly attemptId: string;
  readonly workspaceId: string;
  readonly projectId?: string;
  readonly priority: number;
  /** Deduplication key: the same key must never run twice concurrently. */
  readonly jobKey: string;
  /** ISO-8601. */
  readonly createdAt: string;
  readonly payload: TPayload;
}

/**
 * What the RUNTIME reads off a media payload, before any processor sees it.
 *
 * Deliberately weaker than {@link MediaProbePayload}: `media.acquire` has no
 * source key, so a shared narrow type cannot promise one. Each processor casts
 * to its own payload type, which is where the real shape is asserted.
 */
export interface MediaJobPayload {
  readonly mediaId?: string;
  readonly clipId?: string;
  readonly projectId?: string | null;
  /** Present on the queues that READ an object. */
  readonly key?: string;
  /** Present on the queues that WRITE one. */
  readonly destination?: { readonly key?: string };
}

/**
 * Payload of a `media.probe` job, as `MediaService.startPipeline` writes it.
 *
 * Everything except `mediaId` and `key` is a convenience: the worker rebuilds the
 * CONTRACTS §6 keys from the envelope's ids rather than trusting a path in a
 * payload, so a job replayed from the dead-letter queue a month later still
 * writes to the right place.
 */
export interface MediaProbePayload {
  readonly mediaId: string;
  readonly projectId?: string | null;
  readonly bucket?: string;
  /** Raw object key: `ws/{ws}/p/{project}/media/{media}/raw.{ext}` (CONTRACTS §6). */
  readonly key: string;
  readonly mime?: string | null;
  readonly sizeBytes?: number;
  readonly derivedBucket?: string;
  readonly derivedPrefix?: string;
}

/**
 * Payload of a `media.proxy` job, as the probe's completion handler writes it.
 *
 * The measured fields are hints that save a second probe; `processProxy` measures
 * for itself when one is missing.
 */
export interface MediaProxyPayload extends MediaProbePayload {
  readonly durationMs?: number;
  readonly hasVideo?: boolean;
  readonly hasAudio?: boolean;
  readonly width?: number | null;
  readonly height?: number | null;
  readonly hdr?: boolean;
}

/** Narrow an unknown BullMQ `job.data` to the contract envelope. */
export function isJobEnvelope(value: unknown): value is JobEnvelope {
  if (typeof value !== "object" || value === null) return false;
  const candidate = value as Record<string, unknown>;
  return (
    typeof candidate["jobId"] === "string" &&
    typeof candidate["attemptId"] === "string" &&
    typeof candidate["workspaceId"] === "string" &&
    typeof candidate["jobKey"] === "string" &&
    typeof candidate["createdAt"] === "string" &&
    "payload" in candidate
  );
}

/**
 * Narrow a payload to the minimum every media job needs.
 *
 * Two shapes, because the queues address their object differently and both are
 * legitimate. `media.probe` and `media.proxy` are given a `key` — the object
 * they READ. `media.acquire` has nothing to read yet; it is given
 * `destination.key`, the object it will WRITE. Requiring a top-level `key`
 * rejected every acquisition at the envelope check, which is a permanent failure
 * with a message about CONTRACTS §3 and no hint that the producer and the guard
 * simply disagreed about a field name.
 *
 * What both shapes must have is `mediaId`: the runtime patches that row and
 * reports failures against it before any processor runs.
 */
export function isMediaPayload(value: unknown): value is MediaJobPayload {
  if (typeof value !== "object" || value === null) return false;
  const candidate = value as Record<string, unknown>;
  const hasId =
    (typeof candidate["mediaId"] === "string" && candidate["mediaId"] !== "") ||
    (typeof candidate["clipId"] === "string" && candidate["clipId"] !== "");
  if (!hasId) return false;
  if (typeof candidate["key"] === "string" && candidate["key"] !== "") return true;

  const destination = candidate["destination"];
  if (typeof destination !== "object" || destination === null) return false;
  const key = (destination as Record<string, unknown>)["key"];
  return typeof key === "string" && key !== "";
}

File: apps/worker-ai/worker_ai/queues.py
"""Queue names and the job envelope, frozen in ``docs/CONTRACTS.md`` section 3.

These literals must stay byte-identical to the TypeScript side: the API enqueues
with the Node client and this worker consumes with the Python one, so a typo here
is a silently empty queue rather than a compile error.
``apps/api/src/jobs/contracts/queue-names.test.ts`` parses this file and fails when
the two lists drift.

``montaj`` inside queue names is the engineering codename and is correct — the
brand rule (CONTRACTS section 0) covers user-visible strings only.
"""

from __future__ import annotations

from typing import Any, Final, Literal

from pydantic import BaseModel, ConfigDict, Field, ValidationError

__all__ = [
    "AI_QUEUES",
    "AI_TRANSCRIBE_QUEUE",
    "IMPLEMENTED_AI_QUEUES",
    "QUEUE_NAMES",
    "CompletionStatus",
    "JobEnvelope",
    "bull_job_id",
    "is_job_envelope",
    "parse_envelope",
]

QUEUE_NAMES: Final[tuple[str, ...]] = (
    "media.probe",
    "media.proxy",
    "ai.vad",
    "ai.transcribe",
    "ai.align",
    "ai.diarise",
    "ai.translate",
    "ai.transliterate",
    "ai.clean",
    "ai.pass",
    "ai.llm",
    "render.video",
    "render.subtitle",
    "notify",
)

#: Every queue this worker owns. `media.*`, `render.*` and `notify` belong to the
#: Node workers, so a job on one of them never reaches this process.
AI_QUEUES: Final[tuple[str, ...]] = tuple(name for name in QUEUE_NAMES if name.startswith("ai."))

#: The nine this worker implements (four from A09/A10, `ai.translate` and
#: `ai.transliterate` from A22, `ai.llm` from B11, `ai.pass` from B18,
#: `ai.clean` from B10); the rest are registered and answer `not_implemented`
#: so a producer gets a clear failure instead of a job that sits in Redis
#: forever. `ai.pass` itself still answers `worker/not_implemented` for any
#: `passType` other than `"autocut"` (`processors/autocut_pass.py`) — B19
#: adds `"reframe"`/`"zoom"` there.
IMPLEMENTED_AI_QUEUES: Final[tuple[str, ...]] = (
    "ai.vad",
    "ai.transcribe",
    "ai.align",
    "ai.diarise",
    "ai.translate",
    "ai.transliterate",
    "ai.llm",
    "ai.pass",
    "ai.clean",
)

AI_TRANSCRIBE_QUEUE: Final[str] = "ai.transcribe"

CompletionStatus = Literal["succeeded", "failed"]

_CONTRACT_MESSAGE: Final[str] = "job data does not match the CONTRACTS section 3 envelope"


class JobEnvelope(BaseModel):
    """Every job's ``data``, identical across queues (CONTRACTS section 3).

    Mirrors ``JobEnvelopeSchema`` in ``apps/api/src/jobs/contracts/job-envelope.ts``
    field for field, including the two rules a cross-language consumer depends on:
    every id is a **string**, and ``projectId`` is **omitted** rather than ``null``
    when the job has no project.
    """

    model_config = ConfigDict(populate_by_name=True, frozen=True, extra="ignore")

    job_id: str = Field(alias="jobId", min_length=1)
    attempt_id: str = Field(alias="attemptId", min_length=1)
    workspace_id: str = Field(alias="workspaceId", min_length=1)
    project_id: str | None = Field(default=None, alias="projectId", min_length=1)
    #: BullMQ priority; lower runs first, and the API never sends 0.
    priority: int = Field(default=0, ge=0)
    #: Deduplication key: the same key must never run twice concurrently.
    job_key: str = Field(alias="jobKey", min_length=1)
    #: ISO-8601.
    created_at: str = Field(alias="createdAt", min_length=1)
    payload: dict[str, Any]

    @property
    def bull_job_id(self) -> str:
        """The BullMQ custom id the API allocated: ``{jobId}-{attemptId}``."""
        return bull_job_id(self.job_id, self.attempt_id)

    def log_fields(self) -> dict[str, str]:
        """The identifiers every log line for this job carries."""
        fields = {
            "jobId": self.job_id,
            "attemptId": self.attempt_id,
            "workspaceId": self.workspace_id,
        }
        if self.project_id is not None:
            fields["projectId"] = self.project_id
        return fields


def bull_job_id(job_id: str, attempt_id: str) -> str:
    """BullMQ custom ids may not contain ``:`` — see ``queue.registry.ts``."""
    return f"{job_id}-{attempt_id}"


def is_job_envelope(value: object) -> bool:
    """True when ``value`` carries every field the contract requires."""
    if not isinstance(value, dict):
        return False
    try:
        JobEnvelope.model_validate(value)
    except ValidationError:
        return False
    return True


def parse_envelope(data: object) -> JobEnvelope:
    """Convert BullMQ ``job.data`` into a :class:`JobEnvelope`.

    :raises ValueError: when the job does not match the contract. A malformed job
        is a producer bug, so it fails immediately instead of retrying forever.
    """
    if not isinstance(data, dict):
        raise ValueError(_CONTRACT_MESSAGE)
    try:
        return JobEnvelope.model_validate(data)
    except ValidationError as error:
        fields = ", ".join(str(item["loc"][0]) for item in error.errors() if item["loc"])
        raise ValueError(f"{_CONTRACT_MESSAGE} ({fields})") from error
